New API endpoint: Find 1 user by ID or by email, or list all users.

woutersamaey opened the pull request:
New Greenfield API endpoint to find/list users

Requires the `Unrestricted` permission. I'm open to suggestions if you feel another permission would be preferred.

**List all users**
`GET /api/v1/users`
Returns an array, with same data as the other user calls.

**Find the user with ID xxx**
`GET /api/v1/users?id=xxx`
Returns a single item, with same data as the other user calls.

**Find the user with email abc@example.com**
`GET /api/v1/users?email=abc@example.com`
Returns a single item, with same data as the other user calls.

woutersamaey commented:
I will add to swagger once everyone approves of this call / format.

woutersamaey commented:
Do I need to write a test as well?

woutersamaey commented:
@Kukks How do you feel about a new permission `canmanageusers` ? It would be a breaking change though (for example, we already have the delete a user call and it would apply there as well), so just curious no your thoughts...

bolatovumar commented:
@woutersamaey `CanManageUsers` wouldn't necessarily be a breaking change. It could be a parent policy for `CanDeleteUser` and `CanCreateUser`. So, if a user has `CanManageUsers` permission they can do all of the user stuff. Take a look here: https://github.com/btcpayserver/btcpayserver/blob/e113c1276886c097e061773d6f008c413ec8a464/BTCPayServer.Client/Permissions.cs#L168

NicolasDorier commented on BTCPayServer/Controllers/GreenField/UsersController.cs:
```diff
@@ -64,6 +64,44 @@ public class UsersController : ControllerBase
             _userService = userService;
         }
 
+        [Authorize(Policy = Policies.Unrestricted, AuthenticationSchemes = AuthenticationSchemes.Greenfield)]
```
Shouldn't be Unrestricted by `CanModifyServerSettings`.

woutersamaey replied:
I looked into this and found out we already have `CanDeleteUser` and `CanCreateUser`. In this logic a new permission `CanViewUsers` would make the most sense.
We could add a `CanManageUsers` on top so you can view + create + delete, but in the UI I don't think it would present well as there is no visual queue of permission hierarchy.

So for now I went with `CanViewUsers` as `CanManageUsers` could still be added at a later date.

Also, the values for  `CanDeleteUser` and `CanCreateUser` are a bit poorly selected. 
```
public const string CanCreateUser = "btcpay.server.cancreateuser";
public const string CanDeleteUser = "btcpay.user.candeleteuser";
```
One contains `server` and the other contains `user`. This I cannot fix without BC.

NicolasDorier commented:
see: https://github.com/btcpayserver/btcpayserver/pull/3176/files#r764538054
need a test and documentation in swagger

Blade272008 reviewed (approved):
I need this bad

NicolasDorier commented on BTCPayServer/Controllers/GreenField/UsersController.cs:
```diff
@@ -64,6 +64,44 @@ public class UsersController : ControllerBase
         }
 
+        [Authorize(Policy = Policies.Unrestricted, AuthenticationSchemes = AuthenticationSchemes.Greenfield)]
+        [HttpGet("~/api/v1/users")]
```
it should be two different routes /users and /users/{id}

woutersamaey replied:
Super weird how I missed this. You're absolutely right.

woutersamaey commented:
@NicolasDorier How do I re-generate swagger so I can check the result? Is there a command for that?

woutersamaey commented:
I added a test, but it gives me a 401 unauthorized even though the test client has `Unrestricted` access. I cannot figure it out :(

It's probably something silly, but not seeing it.

NicolasDorier commented:
@woutersamaey what do you mean? Swagger is available to the /docs of your instance. Nothing to regenearate, as the swagger files are hand written.

NicolasDorier commented:
You need to create an admin users and create the client from it, look other tests it does it.

woutersamaey commented:
> @woutersamaey what do you mean? Swagger is available to the /docs of your instance. Nothing to regenearate, as the swagger files are hand written.

I want to (re)generate the HTML output of the Swagger so I can validate the end result.

woutersamaey commented:
> Unrestricted doesn't mean "admin", it means "all permissions of the user". If the user isn't admin, that doesn't work.

Good point. Thanks for the advice!

Kukks commented on BTCPayServer.Client/BTCPayServerClient.Users.cs:
```diff
@@ -26,6 +26,24 @@ public virtual async Task DeleteUser(string userId, CancellationToken token = de
             await HandleResponse(response);
         }
+        
+        public virtual async Task GetUserById(string id, CancellationToken token = default)
```
no return type

woutersamaey replied:
Done

Kukks commented on BTCPayServer.Client/BTCPayServerClient.Users.cs:
```diff
@@ -26,6 +26,24 @@ public virtual async Task DeleteUser(string userId, CancellationToken token = de
+            await HandleResponse(response);
+        }
+        
+        public virtual async Task GetUserByEmail(string email, CancellationToken token = default)
```
no return type.

woutersamaey replied:
Done

Kukks commented on BTCPayServer.Client/BTCPayServerClient.Users.cs:
```diff
@@ -26,6 +26,24 @@ public virtual async Task DeleteUser(string userId, CancellationToken token = de
+            await HandleResponse(response);
+        }
+        
+        public virtual async Task GetUserByEmail(string email, CancellationToken token = default)
```
This seems redundant and should potentially just be bundled in `GetUsers`?

woutersamaey replied:
Done

Kukks commented on BTCPayServer.Client/BTCPayServerClient.Users.cs:
```diff
@@ -26,6 +26,24 @@ public virtual async Task DeleteUser(string userId, CancellationToken token = de
+            await HandleResponse(response);
+        }
+        
+        public virtual async Task GetUsers( CancellationToken token = default)
```
no return type

woutersamaey replied:
Done

Kukks commented on BTCPayServer/Controllers/GreenField/UsersController.cs:
```diff
@@ -64,6 +64,44 @@ public class UsersController : ControllerBase
+        public async Task<ActionResult<ApplicationUserData[]>> GetUsers([FromQuery] string? id = null,
+            [FromQuery] string? email = null)
+        {
+            if (id != null)
```
I do not agree that singular fetching should be bundled in the list fetching endpoint

woutersamaey replied:
Hmm. This is tricky because we can only have 1 route and it should not overlap.

To solve, I named the param `idOrEmail` so you can load by ID or by Email, whichever matches first.

Hope this is ok?

I looked in other places, like invoice loading by orderId, but those calls actually can return multiple results, so it was not a good comparison.

woutersamaey commented:
Thanks for your feedback @Kukks . I'll make the changes at earliest convenience.

Do you have an answer for https://github.com/btcpayserver/btcpayserver/pull/3176#issuecomment-993178687 by any chance?

Kukks commented:
> > @woutersamaey what do you mean? Swagger is available to the /docs of your instance. Nothing to regenearate, as the swagger files are hand written.
> 
> I want to (re)generate the HTML output of the Swagger so I can validate the end result.

There is nothing to generate. The `swagger/v1/swagger.json` file is created dynamically by merging all the files under `BTCPayServer\wwwroot\swagger`

woutersamaey commented:
> 

Maybe I'm explaining myself wrong here, but the docs are HTML generated from `swagger.json`, right? I just want to view the HTML after making changes to the JSON to make sure it looks good in final form. Of course I can just edit the JSON (some areas, not all), but I prefer to see the HTML.

Kukks commented:
> > 
> 
> Maybe I'm explaining myself wrong here, but the docs are HTML generated from `swagger.json`, right? I just want to view the HTML after making changes to the JSON to make sure it looks good in final form. Of course I can just edit the JSON (some areas, not all), but I prefer to see the HTML.

All you have to do is start btcpay and go to `/docs`

Kukks commented:
> > > 
> > 
> > 
> > Maybe I'm explaining myself wrong here, but the docs are HTML generated from `swagger.json`, right? I just want to view the HTML after making changes to the JSON to make sure it looks good in final form. Of course I can just edit the JSON (some areas, not all), but I prefer to see the HTML.
> 
> All you have to do is start btcpay and go to `/docs`

The html loads the json file and renders everything on load from there

woutersamaey commented:
> The html loads the json file and renders everything on load from there

Oh wow! I didn't know this.

woutersamaey commented:
@NicolasDorier I am very close with this one, but the test fails because of some different in handling `System.Net.Http.HttpRequestException` vs `BTCPayServer.Client.GreenFieldAPIException`. Since you made some recent changes, I'm not sure what to do here.

```
Assert.Throws() Failure
Expected: typeof(BTCPayServer.Client.GreenFieldAPIException)
Actual:   typeof(System.Net.Http.HttpRequestException): Response status code does not indicate success: 404 (Not Found).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at BTCPayServer.Client.BTCPayServerClient.HandleResponse(HttpResponseMessage message) in /home/wouter/Business/Workspace/btcpayserver/BTCPayServer.Client/BTCPayServerClient.cs:line 70
   at BTCPayServer.Client.BTCPayServerClient.HandleResponse[T](HttpResponseMessage message) in /home/wouter/Business/Workspace/btcpayserver/BTCPayServer.Client/BTCPayServerClient.cs:line 75
   at BTCPayServer.Client.BTCPayServerClient.GetUserByIdOrEmail(String idOrEmail, CancellationToken token) in /home/wouter/Business/Workspace/btcpayserver/BTCPayServer.Client/BTCPayServerClient.Users.cs:line 33
   at BTCPayServer.Tests.GreenfieldAPITests.<>c__DisplayClass9_0.<<CanViewUsersViaApi>b__3>d.MoveNext() in /home/wouter/Business/Workspace/btcpayserver/BTCPayServer.Tests/GreenfieldAPITests.cs:line 246
```

I have rebased on master already, so it's likely a small detail.

@Kukks I also implemented the changes you requested incl Swagger.

If we can just fix the tests, we should be good to go.

Kukks commented on BTCPayServer/Controllers/GreenField/UsersController.cs:
```diff
@@ -64,6 +64,39 @@ public class UsersController : ControllerBase
+                return Ok(await FromModel(user));
+            }
+
+            return NotFound();
```
```suggestion
            return UserNotFound();
```
this will fix your test

woutersamaey replied:
The return types don't match. How should I fix?
![image](https://user-images.githubusercontent.com/71019/146896368-814a756b-5d3d-4d41-8844-2f8fb8781607.png)

Kukks replied:
Change the return type of the method to `Task<IActionResult>`

woutersamaey replied:
Done.

woutersamaey commented on BTCPayServer/Controllers/GreenField/UsersController.cs:
```diff
@@ -64,6 +64,39 @@ public class UsersController : ControllerBase
+        public async Task<ActionResult<ApplicationUserData[]>> GetUser(string idOrEmail)
+        {
+            // Try ID
+            var user = await _userManager.FindByIdAsync(idOrEmail);
```
Done

woutersamaey commented on BTCPayServer/Controllers/GreenField/UsersController.cs:
```diff
@@ -64,6 +64,39 @@ public class UsersController : ControllerBase
+        [HttpGet("~/api/v1/users/")]
+        public async Task<ActionResult<ApplicationUserData[]>> GetUsers()
+        {
+            var users = _userManager.Users.ToArray();
```
This method does not seam to exist.
![image](https://user-images.githubusercontent.com/71019/146897427-7f7bf369-91d7-4487-9911-3414c830bbbe.png)

woutersamaey replied:
Any ideas @Kukks ? I couldn't figure this out :(

woutersamaey replied:
The method `ToListAsync()` does not exist here, but there is `ToList()`. If I try that, the whole method is no longer async and I get this error.
```
System.InvalidOperationException: A second operation started on this context before a previous operation completed. This is usually caused by different threads using the same instance of DbContext. For more information on how to avoid threading issues with DbContext, see https://go.microsoft.com/fwlink/?linkid=2097913.
   at Microsoft.EntityFrameworkCore.Internal.ConcurrencyDetector.EnterCriticalSection()
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at System.Linq.Enumerable.SelectEnumerableIterator`2.ToList()
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at BTCPayServer.Controllers.GreenField.UsersController.GetUsers() in /home/wouter/Business/Workspace/btcpayserver/BTCPayServer/Controllers/GreenField/UsersController.cs:line 92
   at lambda_method(Closure , Object )
   at Microsoft.Extensions.Internal.ObjectMethodExecutorAwaitable.Awaiter.GetResult()
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ExceptionContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeNextResourceFilter()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at BTCPayServer.Hosting.BTCPayMiddleware.Invoke(HttpContext httpContext) in /home/wouter/Business/Workspace/btcpayserver/BTCPayServer/Hosting/BTCpayMiddleware.cs:line 99
   at BTCPayServer.Hosting.GreenfieldMiddleware.Invoke(HttpContext httpContext) in /home/wouter/Business/Workspace/btcpayserver/BTCPayServer/Hosting/GreenfieldMiddleware.cs:line 30
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at BTCPayServer.Hosting.HeadersOverrideMiddleware.Invoke(HttpContext httpContext) in /home/wouter/Business/Workspace/btcpayserver/BTCPayServer/Hosting/HeadersOverrideMiddleware.cs:line 29
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

Seams I should find some async way of doing the same, but not sure how to do this.

woutersamaey commented:
Almost there. 2 Things remaining.

woutersamaey commented:
@Kukks please help
